### PR TITLE
Ctest and catch integration

### PR DESCRIPTION
**Pull request by ARMartinelli**

### CODE REVIEW

File: tests/CMakeLists.txt [hunk 1]
```diff
@@ -3,12 +3,20 @@ set(BATCH_TESTS_SRCS flatmap.cpp input_output_file.cpp reduce_by_key.cpp
                      read_from_stdin.cpp )
 add_executable(batch_tests ${BATCH_TESTS_SRCS} test_driver.cpp)
 target_link_libraries(batch_tests ${PICO_RUNTIME_LIB})
+target_link_libraries(batch_tests Catch)
+add_test(NAME batch_unit_tests COMMAND batch_tests)
+add_test(NAME stream_unit_tests COMMAND /bin/sh ${CMAKE_CURRENT_SOURCE_DIR}/run_stream_tests.sh)
```
Comment (mdrocco) [inline tests/CMakeLists.txt]: Just for clarity, I would move this line below the block about the `stream_tests` target.

File: tests/CMakeLists.txt [hunk 2]
```diff
@@ -3,12 +3,20 @@ set(BATCH_TESTS_SRCS flatmap.cpp input_output_file.cpp reduce_by_key.cpp
...
 
 set(STREAMING_TESTS_SRCS streaming_reduce_by_key.cpp )
 add_executable(stream_tests ${STREAMING_TESTS_SRCS} test_driver.cpp)
 target_link_libraries(stream_tests ${PICO_RUNTIME_LIB})
 
-add_subdirectory(testdata)
+add_custom_command(TARGET batch_tests
+        POST_BUILD
+        COMMAND /bin/sh ${CMAKE_CURRENT_SOURCE_DIR}/create_inputs_tests.sh
+        )
```
Comment (mdrocco) [inline tests/CMakeLists.txt]: This is dangerous: if `batch_test` target is not built, no input data is generated. Consider also that parallel building could induce a race in input data.

The clean solution would be to treat input data as a target, that would be a dependency for `batch_tests` and `batch_streams` targets.

File: tests/CMakeLists.txt [hunk 3]
```diff
@@ -3,12 +3,20 @@ set(BATCH_TESTS_SRCS flatmap.cpp input_output_file.cpp reduce_by_key.cpp
...
 add_executable(batch_tests ${BATCH_TESTS_SRCS} test_driver.cpp)
 target_link_libraries(batch_tests ${PICO_RUNTIME_LIB})
+target_link_libraries(batch_tests Catch)
+add_test(NAME batch_unit_tests COMMAND batch_tests)
+add_test(NAME stream_unit_tests COMMAND /bin/sh ${CMAKE_CURRENT_SOURCE_DIR}/run_stream_tests.sh)
+
 
 set(STREAMING_TESTS_SRCS streaming_reduce_by_key.cpp )
 add_executable(stream_tests ${STREAMING_TESTS_SRCS} test_driver.cpp)
 target_link_libraries(stream_tests ${PICO_RUNTIME_LIB})
```
Comment (mdrocco) [inline tests/CMakeLists.txt:19]: I think `Catch` should be linked also for `stream_tests`:
```cmake
target_link_libraries(stream_tests Catch)
```

**Review by mdrocco — CHANGES_REQUESTED** [review]

**Comment by mdrocco** [comment]
Please also update the home README.

File: tests/CMakeLists.txt (hunk 1 above)
Comment (ARMartinelli) [reply tests/CMakeLists.txt]: done

File: tests/CMakeLists.txt (hunk 2 above)
Comment (ARMartinelli) [reply tests/CMakeLists.txt]: done

File: tests/CMakeLists.txt (hunk 3 above)
Comment (ARMartinelli) [reply tests/CMakeLists.txt:19]: done

**Review by mdrocco — APPROVED** [review]

**Pull request merged**